File: generated_code/python/solution_297_2.py
import collections

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        # We use a standard Preorder Traversal (DFS) to serialize the tree.
        # This keeps the logic simple and ensures linear O(N) time complexity.
        vals = []
        
        def dfs(node):
            if not node:
                vals.append("#") # Use '#' as a placeholder for null nodes
                return
            vals.append(str(node.val))
            dfs(node.left)
            dfs(node.right)
            
        dfs(root)
        return ",".join(vals) # Join with a delimiter to handle multi-digit/negative numbers

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        # Split the string by the delimiter to get our tokens
        vals = data.split(",")
        # Use an iterator or collections.deque for efficient O(1) popping from the front
        queue = collections.deque(vals)
        
        def dfs():
            if not queue:
                return None
            
            val = queue.popleft()
            if val == "#":
                return None
            
            # Create the current node and reconstruct left and right subtrees recursively
            node = TreeNode(int(val))
            node.left = dfs()
            node.right = dfs()
            return node
            
        return dfs()

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: generated_code/cpp/solution_297_2.cpp
#include <string>
#include <sstream>
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Codec {
private:
    // Helper function for serialization using preorder traversal (DFS).
    void serializeHelper(TreeNode* root, stringstream& ss) {
        if (!root) {
            ss << "# ";
            return;
        }
        // Append the current node's value followed by a space delimiter.
        ss << root->val << " ";
        serializeHelper(root->left, ss);
        serializeHelper(root->right, ss);
    }

    // Helper function for deserialization using a stringstream iterator.
    TreeNode* deserializeHelper(stringstream& ss) {
        string val_str;
        if (!(ss >> val_str)) {
            return nullptr;
        }
        
        // '#' represents a null pointer.
        if (val_str == "#") {
            return nullptr;
        }
        
        // Reconstruct the node and recursively build its left and right subtrees.
        TreeNode* root = new TreeNode(stoi(val_str));
        root->left = deserializeHelper(ss);
        root->right = deserializeHelper(ss);
        return root;
    }

public:
    // Encodes a tree to a single string.
    // Time Complexity: O(N) where N is the number of nodes, as we visit each node once.
    // Space Complexity: O(N) for the recursion stack and the output stream.
    string serialize(TreeNode* root) {
        stringstream ss;
        serializeHelper(root, ss);
        return ss.str();
    }

    // Decodes your encoded data to tree.
    // Time Complexity: O(N) to process the tokens and reconstruct the tree.
    // Space Complexity: O(N) for the call stack during reconstruction and stringstream storage.
    TreeNode* deserialize(string data) {
        if (data.empty()) return nullptr;
        stringstream ss(data);
        return deserializeHelper(ss);
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));